flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/schema/TableSourceTable.scala

```diff
@@ -65,8 +68,28 @@ class TableSourceTable(
 
   override def getQualifiedName: util.List[String] = {
     val builder = ImmutableList.builder[String]()
-        .addAll(super.getQualifiedName)
-    extraDigests.foreach(builder.add)
+      .addAll(super.getQualifiedName)
+
+    if(abilitySpecs != null && abilitySpecs.length != 0){
+//      var newProducedType = catalogTable
+//        .getResolvedSchema
+//        .toSourceRowDataType
+//        .getLogicalType
+//        .asInstanceOf[RowType]
+
+      var newProducedType = DynamicSourceUtils.createProducedType(
+        catalogTable.getResolvedSchema,
+        tableSource)
+
+      for (spec <- abilitySpecs) {
+        val sourceAbilityContext = new SourceAbilityContext(flinkContext, newProducedType)
+
+        builder.add(spec.getDigests(sourceAbilityContext))
+        if (spec.getProducedType.isPresent) {
+          newProducedType = spec.getProducedType.get
+        }
```

Comment:
nit: these lines can be simplified as ` newProducedType = spec.getProducedType.orElse(newProducedType)`